accumulator_updater/programs/accumulator_updater/src/state/accumulator_input.rs

```diff
@@ -14,26 +14,94 @@ use {
 /// the CPI calling program (e.g. Pyth Oracle)
 ///
 /// TODO: implement custom serialization & set alignment
-#[account]
+#[account(zero_copy)]
+#[derive(Debug, InitSpace)]
 pub struct AccumulatorInput {
     pub header: AccumulatorHeader,
-    pub data:   Vec<u8>,
+    // 10KB - 8 (discriminator) - 514 (header)
+    // TODO: do we want to initialize this to the max size?
+    //   - will lead to more data being passed around for validators
+    pub data:   [u8; 9_718],
 }
 
+//TODO:
+// - implement custom serialization & set alignment
+// - what other fields are needed?
+#[zero_copy]
+#[derive(InitSpace, Debug)]
+pub struct AccumulatorHeader {
+    pub bump:        u8, // 1
+    pub version:     u8, // 1
+    // byte offset of accounts where data starts
+    // e.g. account_info.data[offset + header_len]
+    pub header_len:  u16, // 2
+    /// endpoints of every message.
+    /// ex: [10, 14]
+    /// => msg1 = data[(header_len + 0)..(header_len + 10)]
+    /// => msg2 = data[(header_len + 10)..(header_len + 14)]
+    pub end_offsets: [u16; 255], // 510
+}
+
+impl AccumulatorHeader {
+    // header_size. Keeping same structure as `BatchPriceAttestation` header
+    // HEADER_LEN allows for append-only forward-compatibility for the header.
+    // this is the number of bytes from the beginning of the account_info.data
+    // to the start of the `AccumulatorInput` data.
+    //
+    // *NOTE* this implementation is slightly different
+    // than the `BatchPriceAttestation` header_size
+    pub const HEADER_LEN: u16 = 8 + AccumulatorHeader::INIT_SPACE as u16;
+
+    pub const CURRENT_VERSION: u8 = 1;
+
+    pub fn new(bump: u8) -> Self {
+        Self {
+            bump,
+            header_len: Self::HEADER_LEN,
+            version: Self::CURRENT_VERSION,
+            end_offsets: [0u16; u8::MAX as usize],
+        }
+    }
+
+    pub fn set_version(&mut self) {
+        self.version = Self::CURRENT_VERSION;
+    }
+}
 impl AccumulatorInput {
-    pub fn size(data: &Vec<u8>) -> usize {
-        AccumulatorHeader::SIZE + 4 + data.len()
+    pub fn size(&self) -> usize {
+        AccumulatorHeader::INIT_SPACE + 4 + self.data.len()
     }
 
-    pub fn new(header: AccumulatorHeader, data: Vec<u8>) -> Self {
-        Self { header, data }
+    pub fn new(bump: u8) -> Self {
+        let header = AccumulatorHeader::new(bump);
+        Self {
+            header,
+            data: [0u8; 9_718],
+        }
     }
 
-    pub fn update(&mut self, new_data: Vec<u8>) {
-        self.header = AccumulatorHeader::new(self.header.bump, self.header.account_schema);
-        self.data = new_data;
+    // note: this does not handle if multiple CPI calls
+    // need to be made to write all the messages
```

Comment:
This comment is talking about CPIs and stuff which are at a different level of the stack. This comment should describe what this method does literally -- in particular, it overwrites all of the existing data, so you can't append data via multiple calls.

Reply:
also probably good to describe what it returns because it's not self-explanatory